### PR TITLE
[Kernel][Bugfix] Refactor and Fix CUTLASS 2:4 Sparse Kernels

### PR DESCRIPTION
**Pull request by tlrmchlsmth**
This PR generally reworks the 2:4 sparse CUTLASS kernels, cleaning up the implementation, and fixes several issues.

* `cutlass_sparse_compress` and `cutlass_scaled_sparse_mm` now share the dispatching for hyperparameter selection. This lets the compressor derive the sparse format from the GEMM kernel. Now the sparse format has a single source of truth and it lets us pare down the code size for `cutlass_sparse_compress`.
* `cutlass_sparse_compress` now allocates and returns `a_nzs` and `a_meta`. Necessary because `cutlass_sparse_compress` requires padding in some cases (e.g. where the K dimension is not a multiple of 128).
* Fixes an IMA and/or garbage output when `bias is not None`. `cutlass_scaled_sparse_mm` internally transposes the operation, which means the bias is a column vector instead of a row vector.  Added a separate epilogue for that case. (this happened in the qkv_proj in `nm-testing/Qwen2.5-0.5B-sparse2of4_fp8_dynamic-e2e`)
* Fixes an issue where float16 or bfloat16 2:4 models where zero entries were stored as `-0.0` would produce garbage output (`nm-testing/llama-2-7b-sparse2of4`).

**Note**: With this PR I updated the kernels so that bfloat16 and float16 `cutlass_scaled_sparse_mm` GEMMs must have the same output dtype as their input operands. Purely for binary size/compile time reasons. Easy to add back in if anyone has an objection.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -228,7 +228,7 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
   SET(CUTLASS_ENABLE_HEADERS_ONLY ON CACHE BOOL "Enable only the header library")
 
   # Set CUTLASS_REVISION manually -- its revision detection doesn't work in this case.
-  set(CUTLASS_REVISION "v3.6.0" CACHE STRING "CUTLASS revision to use")
+  set(CUTLASS_REVISION "v3.7.0" CACHE STRING "CUTLASS revision to use")
```
Comment (alexm-redhat) [inline CMakeLists.txt:231]: why it was necessary to update to 3.7?

File: csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c2x.hpp [hunk 2]
```diff
@@ -316,4 +316,4 @@ struct ScaledEpilogueBiasAzpToken
   }
 };
 
-};  // namespace vllm::c2x
\ No newline at end of file
+};  // namespace vllm::c2x
```
Comment (alexm-redhat) [inline csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c2x.hpp:319]: formatting?

File: csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c3x.hpp [hunk 3]
```diff
@@ -174,6 +198,49 @@ struct ScaledEpilogueBias
   }
 };
 
+/*
+ * This epilogue performs the same operation as ScaledEpilogueBias, but the
+ * bias, but the bias is a column vector. Useful e.g. if we are computing a
```
Comment (alexm-redhat) [inline csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c3x.hpp:203]: but the bias => duplicate text

File: csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c3x.hpp [hunk 4]
```diff
@@ -314,4 +381,4 @@ struct ScaledEpilogueBiasAzpToken
   }
 };
 
-};  // namespace vllm::c3x
\ No newline at end of file
+};  // namespace vllm::c3x
```
Comment (alexm-redhat) [inline csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c3x.hpp:384]: formatting

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm.cuh [hunk 5]
```diff
@@ -53,12 +53,15 @@ struct cutlass_3x_gemm {
 
   using EVTCompute = typename Epilogue::EVTCompute;
 
+  static constexpr int AlignmentAB = 16;
```
Comment (alexm-redhat) [inline csrc/quantization/cutlass_w8a8/c3x/scaled_mm.cuh:56]: A brief comment of why 16 and 4 would be helpful

File: csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cuh [hunk 6]
```diff
@@ -103,14 +103,17 @@ struct cutlass_2x_gemm {
 
   using EVTD = cutlass::epilogue::threadblock::Sm80EVT<D, EVTCompute>;
 
+  static constexpr int AlignmentAB = 16;
```
Comment (alexm-redhat) [inline csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cuh:106]: ditto

File: csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu [hunk 7]
```diff
@@ -49,122 +62,87 @@ void cutlass_gemm_sm90_fp8_dispatch(torch::Tensor& out, torch::Tensor const& a,
   using Cutlass3xGemm8 =
       typename sm90_fp8_config_8<InType, OutType, Epilogue>::Cutlass3xGemm;
 
-  uint32_t const n = bt_nzs.size(0);
-  uint32_t const m = a.size(0);  // Batch size
   uint32_t const mp2 =
       std::max(static_cast<uint32_t>(64), next_pow_2(m));  // next power of 2
 
   if (mp2 <= 64) {
     if (n == 28672) {
```
Comment (alexm-redhat) [inline csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu:69]: 28672 seems like a magic number, a comment would be helpful

File: csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu [hunk 8]
```diff
@@ -49,122 +62,87 @@ void cutlass_gemm_sm90_fp8_dispatch(torch::Tensor& out, torch::Tensor const& a,
...
+    return DispatchFunc::template invoke<Cutlass3xGemmM128>(
+        std::forward<Args>(args)...);
   } else if (mp2 <= 256) {
     // n in (128, 256]
-    return cutlass_sparse_gemm_caller<Cutlass3xGemmM256>(
-        out, a, bt_nzs, bt_meta, std::forward<EpilogueArgs>(args)...);
+    return DispatchFunc::template invoke<Cutlass3xGemmM256>(
+        std::forward<Args>(args)...);
   } else {
     // n in (256, inf)
```
Comment (alexm-redhat) [inline csrc/sparse/cutlass/sparse_scaled_mm_c3x.cu:122]: the dispatching mechanism is hardcoded to specific numbers, does it work for different GPUs? or it is mainly for H100?

File: csrc/sparse/cutlass/sparse_scaled_mm_entry.cu [hunk 9]
```diff
@@ -9,7 +9,7 @@ bool cutlass_sparse_scaled_mm_supported(int64_t cuda_device_capability) {
   // sparse CUTLASS kernels need at least
   //   CUDA 12.2 and SM90 (Hopper)
 
-#if defined CUDA_VERSION
+#if defined ENABLE_SPARSE_SCALED_MM_C3X && ENABLE_SPARSE_SCALED_MM_C3X
```
Comment (alexm-redhat) [inline csrc/sparse/cutlass/sparse_scaled_mm_entry.cu:12]: A comment would be helpful, non trivial if def.

File: tests/kernels/test_cutlass_2of4_sparse.py [hunk 10]
```diff
@@ -55,11 +54,33 @@ def prune_to_2_4(tensor):
     return pruned.reshape(original_shape)
 
 
+# This function checks that applying an identity matrix multiplication
+# to the compressed weights yields the original uncompressed weights.
+def check_compress_decompress_invariance(dtype: torch.dtype, b: torch.Tensor,
+                                         b_compressed: torch.Tensor,
+                                         b_metadata: torch.Tensor):
+    out_dtype = torch.float16 if dtype is torch.float16 else torch.bfloat16
```
Comment (alexm-redhat) [inline tests/kernels/test_cutlass_2of4_sparse.py:62]: bfloat16 as default is something expected? or we want the dtype to be explicit always?

File: tests/kernels/test_cutlass_2of4_sparse.py [hunk 11]
```diff
@@ -55,11 +54,33 @@ def prune_to_2_4(tensor):
...
+                                            eye_scale,
+                                            out_dtype=out_dtype)
+
+    torch.testing.assert_close(b.to(dtype=out_dtype), b_decomp)
+
+
 def make_rand_sparse_tensors(
         dtype: torch.dtype, m: int, n: int, k: int
 ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
-    a = torch.randn((m, k), device='cuda') * 5
```
Comment (alexm-redhat) [inline tests/kernels/test_cutlass_2of4_sparse.py:61]: why *5?

File: tests/kernels/test_cutlass_2of4_sparse.py [hunk 12]
```diff
@@ -134,53 +156,70 @@ def test_cutlass_sparse_subset():
...
 
-    torch.testing.assert_close(out, baseline, rtol=1e-2, atol=1e-2)
+    baseline = baseline_scaled_mm(a,
+                                  b,
+                                  scale_a,
+                                  scale_b,
+                                  out_dtype=dtype,
+                                  bias=bias)
+
+    torch.testing.assert_close(out, baseline, rtol=1e-2, atol=3e-1)
```
Comment (alexm-redhat) [inline tests/kernels/test_cutlass_2of4_sparse.py:189]: nice tests

File: tests/kernels/test_cutlass_2of4_sparse.py [hunk 13]
```diff
@@ -134,53 +156,70 @@ def test_cutlass_sparse_subset():
...
 
     baseline = baseline_scaled_mm(a,
                                   b,
                                   scale_a,
                                   scale_b,
-                                  out_dtype=torch.bfloat16)
+                                  out_dtype=out_dtype,
+                                  bias=bias)
 
     torch.testing.assert_close(out, baseline, rtol=1e0, atol=2e0)
```
Comment (alexm-redhat) [inline tests/kernels/test_cutlass_2of4_sparse.py:224]: 1e0 and 2e0 seem to be pretty large, no?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py [hunk 14]
```diff
@@ -205,6 +204,11 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
                 layer.weight_scale = torch.nn.Parameter(
                     layer.weight_scale.data, requires_grad=False)
 
+        # Set all negative zero values to 0 prior to compression
+        if (layer.weight.dtype.is_floating_point
+                and layer.weight.dtype.itemsize >= 2):
+            layer.weight.data[layer.weight.data == -0.0] = 0.0
```
Comment (alexm-redhat) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:210]: nice catch, never though that a compiler will be sensitive to -0 or +0 :)

**Review by alexm-redhat — COMMENTED** [review]
LGTM in general, I don't fully understand the template of CUTLASS, but the code makes sense.

File: CMakeLists.txt (hunk 1 above)
Comment (tlrmchlsmth) [reply CMakeLists.txt:231]: We're already on 3.7. I just noticed this is out of date, which just means some incorrect messages are printed while building the cutlass kernels. Going to leave some reminder comments

File: csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c3x.hpp (hunk 4 above)
Comment (tlrmchlsmth) [reply csrc/cutlass_extensions/epilogue/scaled_mm_epilogues_c3x.hpp:384]: My editor adds this by default

File: tests/kernels/test_cutlass_2of4_sparse.py (hunk 10 above)
Comment (tlrmchlsmth) [reply tests/kernels/test_cutlass_2of4_sparse.py:62]: I think the code is fine but is kind of weird/hacky/confusing, so I left a comment.

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py [hunk 15]
```diff
@@ -254,9 +258,10 @@ def apply_weights(
             bt_meta=layer.meta,
             scale_a=input_scale,
             scale_b=layer.weight_scale,
-            out_dtype=self.output_dtype,
+            out_dtype=x.dtype,
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:261]: We should run some of the cases for each of tests here onwards: https://github.com/neuralmagic/vllm/blob/4c0d93f4b2de241336f4732cb5799cee8fedcb52/tests/quantization/test_compressed_tensors.py#L305

File: tests/kernels/test_cutlass_2of4_sparse.py (hunk 11 above)
Comment (tlrmchlsmth) [reply tests/kernels/test_cutlass_2of4_sparse.py:61]: I added a comment below, also changed this to only happen for the int8 case (where it's needed so the inputs aren't all 0s)

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py (hunk 15 above)
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:261]: Good call. I ran this locally on an H100 and it passes:
`pytest tests/quantization/test_compressed_tensors.py`

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm.cuh (hunk 5 above)
Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/c3x/scaled_mm.cuh]: Updated.

**Review by alexm-redhat — APPROVED** [review]
LGTM
